File: samples/ruby/googletests.rb
require 'rubygems'
require 'test/unit'
require 'appium_lib'

class AndroidTests < Test::Unit::TestCase
  USER_NAME = 'tester01'
  API_KEY = '13e36639-92e3-411b-a067-3457b5dea573'
  HOST_NAME = 'ec2-54-226-177-179.compute-1.amazonaws.com'
  PORT = '3001'
  GOOGLE_URL = 'https://mail.google.com'
  GOOGLE_URL_HOMEPAGE = 'https://mail.google.com/mail/mu/mp/4/#tl/priority/%5Esmartlabel_personal'
  TRASH_URL = 'https://mail.google.com/mail/mu/mp/4/#tl/Trash'
  SENT_EMAIL_URL='https://mail.google.com/mail/mu/mp/68/#tl/Sent%20Mail'
  PRIMARY_EMAIL_URL='https://mail.google.com/mail/mu/mp/68/#tl/priority/%5Esmartlabel_personal'
  EMAIL_ADDRESS1 = 'krypton.portal@gmail.com'
  EMAIL_ADDRESS2 = 'krypton.portal2@gmail.com'
  PASSWORD = 'Admin@123456'
  SUBJECT1 = 'It is a subject'
  BODY1 = 'It is a body'
  MSG_NOEMAIL = 'You have no mail here.'
  MSG_NOEMAIL_PRIMARY1 = 'You have no mail.'
  MSG_NOEMAIL_PRIMARY2 = 'Please enjoy your day!'
  SERVER_URL_LOCAL='http://127.0.0.1:4723/wd/hub'

  def define_url
    url = "http://" + USER_NAME + ":" + API_KEY + "@" + HOST_NAME + ":" + PORT + "/v1/tests"
    return url
  end

    def setup
      #puts(define_url)
        desired_caps = {
            caps:       {
            'browserName'=> 'Chrome',
            'platformVersion' => '5.1.1',
            'platformName' => 'Android',
            'deviceName' => 'Nexus 5'
            },
            appium_lib: {
                server_url: SERVER_URL_LOCAL
            }
        }
        @driver = Appium::Driver.new(desired_caps)
        @driver.start_driver.manage.timeouts.implicit_wait = 30
        @driver.driver.manage.timeouts.page_load = 30
        @driver.driver.manage.delete_all_cookies
    end

    def teardown
        unless @driver.driver.nil?
          @driver.driver.quit
          sleep(10)
        end
    end

  def test_should_not_accept_empty_or_invalid_email
      @driver.driver.navigate.to(GOOGLE_URL)
      sleep(3)
      @driver.find_element(:id, 'next').click
      sleep(3)
      getText = @driver.find_element(:id, 'errormsg_0_Email').text()
      assert_equal(getText, 'Please enter your email.')
      @driver.find_element(:id, 'Email').send_keys 'invalid_email@where.about'
      @driver.find_element(:id, 'next').click
      sleep(3)
      getText = @driver.find_element(:id, 'errormsg_0_Email').text()
      assert_equal(getText, "Sorry, Google doesn't recognize that email. Create an account using that address?")
  end

  def test_should_accept_valid_credentials
      @driver.driver.navigate.to(GOOGLE_URL)
      sleep(3)
      @driver.find_element(:id, 'Email').send_keys EMAIL_ADDRESS1
      @driver.find_element(:id, 'next').click
      sleep(1)
      @driver.find_element(:id, 'Passwd').send_keys PASSWORD
      @driver.find_element(:id, 'signIn').click
      sleep(10)
      getURL=@driver.driver.current_url
      assert_equal(true,(getURL.include? 'https://mail.google.com/mail'))
  end

  def test_should_compose_email_successfully_by_Gmail1
      login EMAIL_ADDRESS1, PASSWORD
      compose_email EMAIL_ADDRESS2, SUBJECT1, BODY1
      verify_sent_email SUBJECT1, BODY1, EMAIL_ADDRESS1, EMAIL_ADDRESS2
      delete_sent_email
      delete_trash_email
      logout
      login EMAIL_ADDRESS2, PASSWORD
      verify_primary_email SUBJECT1, BODY1, EMAIL_ADDRESS1, EMAIL_ADDRESS2
      delete_primary_email
      delete_trash_email
  end

  def login(username, password)
    @driver.driver.navigate.to(GOOGLE_URL)
    sleep(3)
    @driver.find_element(:id, 'Email').send_keys username
    @driver.find_element(:id, 'next').click
    sleep(1)
    @driver.find_element(:id, 'Passwd').send_keys password
    @driver.find_element(:id, 'signIn').click
    sleep(10)
  end

  def logout
    @driver.driver.navigate.to(PRIMARY_EMAIL_URL)
    sleep(3)
    @driver.find_element(:xpath, "//div[@aria-label='Menu']").click
    sleep(1)
    @driver.find_element(:xpath, "//div[contains(@class,'V Y Rx Kg')]").click
    sleep(1)
    @driver.find_element(:xpath, "//button[@id='signout']").click
    sleep(5)
    @driver.driver.navigate.to(GOOGLE_URL)
    @driver.driver.manage.delete_all_cookies
    @driver.driver.navigate.refresh
    sleep(5)
  end

  def compose_email(toEmail, subject, body)
    @driver.find_element(:xpath, "//div[@id='views']//div[@aria-label='Compose']").click
    sleep(1)
    @driver.find_element(:xpath, "//div[@id='cmcc_composeto']//input[@id='composeto']").send_keys toEmail
    @driver.find_element(:xpath, "//input[@id='cmcsubj']").send_keys subject
    @driver.find_element(:id, "cmcbody").send_keys body
    sleep(3)
    @driver.find_element(:xpath, "//div[@id='views']//div[text()='Send']").click
    sleep(1)
  end

  def verify_sent_email(subject, body, sendEmail, toEmail)
    @driver.find_element(:xpath, "//div[@aria-label='Menu']").click
    sleep(3)
    @driver.find_element(:xpath, "//span[text()='Sent Mail']").click
    sleep(3)
    getText = @driver.find_element(:xpath, "//div[@class='fm']").text()
    assert_equal(true,(getText.include? subject))
    assert_equal(true,(getText.include? body))
    @driver.find_element(:xpath, "//div[contains(@class,'m')]//div[@role='listitem'][1]").click
    sleep(1)
    @driver.find_element(:xpath, "//div[contains(@class,'V j hj') and text()='Details']").click
    getSubject = @driver.find_element(:xpath, "//span[@class='kj']/span").text()
    assert_equal(true,(getSubject.include? subject))
    getBody = @driver.find_element(:xpath, "//div[@class='Hi']").text()
    assert_equal(true,(getBody.include? body))
    getSendEmail = @driver.find_element(:xpath, "//div[@class='Kg']/span").text()
    assert_equal(true,(getSendEmail.include? sendEmail))
    getToEmail = @driver.find_element(:xpath, "//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']").text()
    assert_equal(true,(getToEmail.include? toEmail))
  end

  def verify_primary_email(subject, body, sendEmail, toEmail)
    @driver.driver.navigate.to(GOOGLE_URL_HOMEPAGE)
    @driver.driver.navigate.refresh
    sleep(3)
    getText = @driver.find_element(:xpath, "//div[@class='fm']").text()
    assert_equal(true,(getText.include? subject))
    assert_equal(true,(getText.include? body))
    @driver.find_element(:xpath, "//div[contains(@class,'fm')]//div[@role='listitem'][1]").click
    sleep(5)
    @driver.find_element(:xpath, "//div[contains(@class,'V j hj') and text()='Details']").click
    getSubject = @driver.find_element(:xpath, "//span[@class='kj']/span").text()
    assert_equal(true,(getSubject.include? subject))
    getBody = @driver.find_element(:xpath, "//div[@class='Hi']").text()
    assert_equal(true,(getBody.include? body))
    getSendEmail = @driver.find_element(:xpath, "//div[@class='Kg']/span").text()
    assert_equal(true,(getSendEmail.include? sendEmail))
    getToEmail = @driver.find_element(:xpath, "//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']").text()
    assert_equal(true,(getToEmail.include? toEmail))
  end

  def delete_sent_email
    @driver.driver.navigate.to(SENT_EMAIL_URL)
    sleep(2)
    @driver.find_element(:xpath, "//div[@class='V j cb Ol'][1]").click
    sleep(1)
    @driver.find_element(:xpath, "//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']").click
    sleep(1)
    noEmailSentEmail = @driver.find_element(:xpath, "//div[@class='Wl']").text()
    assert_equal(noEmailSentEmail, MSG_NOEMAIL)
  end

  def delete_trash_email
    @driver.driver.navigate.to(TRASH_URL)
    sleep(2)
    @driver.find_element(:xpath, "//div[@class='V j cb Ol'][1]").click
    sleep(1)
    @driver.find_element(:xpath, "//div[@class='M j T b hc q m']//div[text()='Delete forever']").click
    sleep(1)
    noEmailTrash = @driver.find_element(:xpath, "//div[@class='Wl']").text()
    assert_equal(noEmailTrash, MSG_NOEMAIL)
  end

  def delete_primary_email
    @driver.driver.navigate.to(PRIMARY_EMAIL_URL)
    sleep(2)
    @driver.find_element(:xpath, "//div[@class='V j cb Ol'][1]").click
    sleep(1)
    @driver.find_element(:xpath, "//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']").click
    sleep(1)
    noEmailPrimary = @driver.find_element(:xpath, "//div[@class='Wl']").text()
    assert_equal(true,(noEmailPrimary.include? MSG_NOEMAIL_PRIMARY1))
    assert_equal(true,(noEmailPrimary.include? MSG_NOEMAIL_PRIMARY2))
  end
end

File: samples/java/java/src/main/java/google/GoogleTests.java
package google;

import java.net.MalformedURLException;
import java.net.URL;

import org.apache.commons.codec.binary.Base64;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.CapabilityType;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.internal.ApacheHttpClient;
import org.testng.Assert;
import org.testng.annotations.AfterTest;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Parameters;
import org.testng.annotations.Test;

import io.appium.java_client.android.AndroidDriver;
import io.appium.java_client.remote.MobileCapabilityType;
import io.appium.java_client.remote.MobilePlatform;

public class GoogleTests {
	private AndroidDriver<WebElement> driver = null;
	URL url = null;
	String USER_NAME = "tester01";
	String API_KEY = "fe82cbb9-a8df-4c60-95cc-90a2ac1ccf21";
	String HOST_NAME = "ec2-54-226-177-179.compute-1.amazonaws.com";
	String PORT = "3001";
	String GOOGLE_URL = "https://mail.google.com";
	String GOOGLE_URL_HOMEPAGE = "https://mail.google.com/mail/mu/mp/4/#tl/priority/%5Esmartlabel_personal";
	String TRASH_URL = "https://mail.google.com/mail/mu/mp/4/#tl/Trash";
	String EMAIL_ADDRESS1 = "krypton.portal@gmail.com";
	String EMAIL_ADDRESS2 = "krypton.portal2@gmail.com";
	String PASSWORDSEND = "Admin@123456";
	String PASSWORDTO = "Admin@123456";
	String SUBJECT1 = "It is a subject";
	String BODY1 = "It is a body";
	String MSG_NOEMAIL = "You have no mail here.";
	String MSG_NOEMAIL_PRIMARY = "You have no mail.\nPlease enjoy your day!";

	@Parameters({ "browserName", "deviceName", "platformVersion", "emailSend", "passwordSend", "emailTo", "passwordTo", "username", "apikey" })
	@BeforeTest
	public void Setup(String browserName, String deviceName, String platformVersion, String emailSend, String passwordSend, String emailTo,
			String passwordTo, String username, String apikey) {

		EMAIL_ADDRESS1 = emailSend;
		PASSWORDSEND = passwordSend;
		EMAIL_ADDRESS2 = emailTo;
		PASSWORDTO = passwordTo;
		USER_NAME= username;
		API_KEY=apikey;
		DesiredCapabilities capabilities = DesiredCapabilities.android();
		capabilities.setCapability(MobileCapabilityType.BROWSER_NAME, browserName);
		capabilities.setCapability(MobileCapabilityType.PLATFORM_NAME, MobilePlatform.ANDROID);
		capabilities.setCapability(MobileCapabilityType.DEVICE_NAME, deviceName);
		capabilities.setCapability(MobileCapabilityType.PLATFORM_VERSION, platformVersion);
		capabilities.setCapability(CapabilityType.ACCEPT_SSL_CERTS, true);

		try {
			String parseUrl="http://" + USER_NAME +":" + API_KEY +"@" +  HOST_NAME +":"+PORT+"/wd/hub";
			url = new URL(parseUrl);
		} catch (MalformedURLException e) {
			e.printStackTrace();
		}
		
		driver = new AndroidDriver<WebElement>(url, capabilities);
		driver.manage().deleteAllCookies();
		driver.navigate().refresh();
	}

	@AfterTest
	public void Teardown() {
		try {
			driver.quit();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Test(priority = 1, description = "should not accept empty email or invalid email")
	public void GoogleLogin_Invalid() {
		driver.get(GOOGLE_URL);
		Sleep(5000);
		driver.findElementById("next").click();
		Sleep(2000);
		String errorMsg = driver.findElementById("errormsg_0_Email").getText();
		Assert.assertEquals(errorMsg, "Please enter your email.");
		driver.findElementById("Email").sendKeys("invalid_email@where.about");
		driver.findElementById("next").click();
		Sleep(2000);
		String errorMsg2 = driver.findElementById("errormsg_0_Email").getText();
		Assert.assertEquals(errorMsg2,
				"Sorry, Google doesn't recognize that email. Create an account using that address?");
	}

	@Test(priority = 2, description = "should accept valid credentials")
	public void GoogleLogin_Valid() {
		Login(EMAIL_ADDRESS1, PASSWORDSEND);
		Sleep(5000);
		String getURL = driver.getCurrentUrl();
		Assert.assertEquals(true, getURL.contains("https://mail.google.com/mail"));
	}

	@Test(priority = 3, description = "should compose email successfully by Gmail1")
	public void GoogleComposeEmail() {
		// should compose email successfully by Gmail1
		driver.findElementByXPath("//div[@id='views']//div[@aria-label='Compose']").click();
		Sleep(1000);
		driver.findElementByXPath("//div[@id='cmcc_composeto']//input[@id='composeto']").sendKeys(EMAIL_ADDRESS2);
		driver.findElementByXPath("//input[@id='cmcsubj']").sendKeys(SUBJECT1);
		driver.findElementById("cmcbody").sendKeys(BODY1);
		Sleep(1000);
		driver.findElementByXPath("//div[@id='views']//div[text()='Send']").click();
		Sleep(1000);
		driver.findElementByXPath("//div[@aria-label='Menu']").click();
		Sleep(1000);
		driver.findElementByXPath("//span[text()='Sent Mail']").click();
		Sleep(1000);

		// verify the new email exists on Sent Email folder on Gmail1
		String getText = driver.findElementByXPath("//div[@class='fm']").getText();
		Assert.assertEquals(true, getText.contains(SUBJECT1));
		Assert.assertEquals(true, getText.contains(BODY1));
		driver.findElementByXPath("//div[contains(@class,'m')]//div[@role='listitem'][1]").click();
		Sleep(2000);
		driver.findElementByXPath("//div[contains(@class,'V j hj') and text()='Details']").click();
		Sleep(3000);
		String getSubject = driver.findElementByXPath("//span[@class='kj']/span").getText();
		String getBody = driver.findElementByXPath("//div[@class='Hi']").getText();
		String getSentEmail = driver.findElementByXPath("//div[@class='Kg']/span").getText();
		String getToEmail = driver.findElementByXPath("//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']")
				.getText();

		Assert.assertEquals(SUBJECT1, getSubject);
		Assert.assertEquals(BODY1, getBody);
		Assert.assertEquals(EMAIL_ADDRESS1, getSentEmail);
		Assert.assertEquals(EMAIL_ADDRESS2, getToEmail);

		// delete email on Sent Email and Trash folder on Gmail1
		driver.findElementByXPath("//div[@id='cv__cntbt']//div[@class='V j Y Mm Kg']").click();
		Sleep(1000);
		driver.findElementByXPath("//div[@class='V j cb Ol'][1]").click();
		Sleep(1000);
		driver.findElementByXPath("//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']").click();
		Sleep(2000);
		String getNoEmailMsg = driver.findElementByXPath("//div[@class='Wl']").getText();

		Assert.assertEquals(MSG_NOEMAIL, getNoEmailMsg);

		driver.get(TRASH_URL);
		Sleep(5000);
		driver.findElementByXPath("//div[@class='V j cb Ol'][1]").click();
		Sleep(1000);
		driver.findElementByXPath("//div[@class='M j T b hc q m']//div[text()='Delete forever']").click();
		getNoEmailMsg = driver.findElementByXPath("//div[@class='Wl']").getText();

		Assert.assertEquals(MSG_NOEMAIL, getNoEmailMsg);

		driver.findElementByXPath("//div[@aria-label='Menu']").click();
		Sleep(1000);
		driver.findElementByXPath("//div[contains(@class,'V Y Rx Kg')]").click();
		Sleep(1000);
		driver.findElementByXPath("//button[@id='signout']").click();
		Sleep(5000);

		driver.get(GOOGLE_URL);
		driver.manage().deleteAllCookies();
		driver.navigate().refresh();

		// verify Gmail2 received a new email from Gmail1
		Login(EMAIL_ADDRESS2, PASSWORDTO);
		driver.get(GOOGLE_URL_HOMEPAGE);
		driver.navigate().refresh();
		Sleep(5000);

		getText = driver.findElementByXPath("//div[@class='fm']").getText();
		Assert.assertEquals(true, getText.contains(SUBJECT1));
		Assert.assertEquals(true, getText.contains(BODY1));
		driver.findElementByXPath("//div[contains(@class,'fm')]//div[@role='listitem'][1]").click();
		Sleep(1000);
		driver.findElementByXPath("//div[contains(@class,'V j hj') and text()='Details']").click();
		Sleep(2000);
		getSubject = driver.findElementByXPath("//span[@class='kj']/span").getText();
		getBody = driver.findElementByXPath("//div[@class='Hi']").getText();
		getSentEmail = driver.findElementByXPath("//div[@class='Kg']/span").getText();
		getToEmail = driver.findElementByXPath("//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']").getText();

		Assert.assertEquals(SUBJECT1, getSubject);
		Assert.assertEquals(BODY1, getBody);
		Assert.assertEquals(EMAIL_ADDRESS1, getSentEmail);
		Assert.assertEquals(EMAIL_ADDRESS2, getToEmail);

		// delete email on Primary Email and Trash folder on Gmail2
		driver.findElementByXPath("//div[contains(@class,'M j T b hc Om o')]//div[text()='Primary'][1]").click();
		Sleep(1000);
		driver.findElementByXPath("//div[@class='V j cb Ol'][1]").click();
		Sleep(1000);
		driver.findElementByXPath("//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']").click();
		getNoEmailMsg = driver.findElementByXPath("//div[@class='Wl']").getText();

		Assert.assertEquals(MSG_NOEMAIL_PRIMARY, getNoEmailMsg);

		driver.get(TRASH_URL);
		Sleep(5000);
		driver.findElementByXPath("//div[@class='V j cb Ol'][1]").click();
		Sleep(1000);
		driver.findElementByXPath("//div[@class='M j T b hc q m']//div[text()='Delete forever']").click();
		getNoEmailMsg = driver.findElementByXPath("//div[@class='Wl']").getText();

		Assert.assertEquals(MSG_NOEMAIL, getNoEmailMsg);
	}

	private void Login(String username, String password) {
		driver.get(GOOGLE_URL);
		Sleep(5000);
		driver.findElementById("Email").sendKeys(username);
		driver.findElementById("next").click();
		Sleep(1000);
		driver.findElementById("Passwd").sendKeys(password);
		driver.findElementById("signIn").click();
		Sleep(10000);
	}

	private void Sleep(int millis) {
		try {
			Thread.sleep(millis);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}

}
